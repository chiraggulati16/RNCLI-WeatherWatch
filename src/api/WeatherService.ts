import axios from "axios";
import { Location } from "../models/Location";

const BASE_URL = 'https://geocoding-api.open-meteo.com/v1/';
const WEATHER_BASE_URL = "https://api.open-meteo.com/v1/";
export class WeatherService {
    static async getCurrentWeather(location: Location) {
      const response = await axios.get(
        `${WEATHER_BASE_URL}forecast?latitude=${location.latitude}&longitude=${location.longitude}&current=temperature_2m,weather_code&hourly=temperature_2m,weather_code`
      );
      console.log("result 11", response.data)
      return response.data;
    }
  
    static async getSearchList(location: string) {
      const response = await axios.get(
        `${BASE_URL}search?name=${location}`
      );
      return response.data.results;
    }
  }