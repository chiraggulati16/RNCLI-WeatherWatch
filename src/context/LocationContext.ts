import React, {createContext, useContext} from 'react';
import {Location} from '../models/Location';

// Define a type for the context value
export type AppContextType = {
  location: Location;
  setLocation: React.Dispatch<React.SetStateAction<Location>>;
};

// Create the context with a default value (optional)
export const LocationContext = createContext<AppContextType | undefined>(
  undefined,
);

// Custom hook to use the AppContext
export const useLocationContext = (): AppContextType => {
  const context = useContext(LocationContext);
  if (!context) {
    throw new Error('useAppContext must be used within an AppProvider');
  }
  return context;
};
