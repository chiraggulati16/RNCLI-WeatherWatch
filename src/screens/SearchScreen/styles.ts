import {StyleSheet} from 'react-native';

export const styles = StyleSheet.create({
  container: {padding: 16, flex: 1},
  item: {padding: 8, fontSize: 18, color: '#000'},
});
