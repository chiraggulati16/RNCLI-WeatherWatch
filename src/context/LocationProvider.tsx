import { ReactNode, useState } from "react";
import { LocationContext } from "./LocationContext";
import { Location } from "../models/Location";

interface MyProviderProps {
    children: ReactNode;
  }
  
  export const LocationProvider: React.FC<MyProviderProps> = ({ children }) => {
    const [location, setLocation] = useState<Location>({
        id: 1,
        name: "Chandigarh",
        latitude: 30.7333,
        longitude: 76.7794,
        country: "India"
      })
  
    return (
      <LocationContext.Provider value={{ location, setLocation }}>
        {children}
      </LocationContext.Provider>
    );
  };