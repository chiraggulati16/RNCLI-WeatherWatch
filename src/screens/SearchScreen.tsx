import {FlatList, StyleSheet, Text, TouchableOpacity, View} from 'react-native';
import {WeatherService} from '../api/WeatherService';
import {Location} from '../models/Location';
import SearchBar from '../components/SearchBar';
import { useEffect, useState} from 'react';
import { useLocationContext } from '../context/LocationContext';
import NavigationService from '../navigation/NavigationService';

const SearchScreen = () => {
  const [locations, setLocations] = useState<Location[]>([]);
  const {setLocation} = useLocationContext();

  const handleSearch = async (location: string) => {
    console.log("result before")
    const results = await WeatherService.getSearchList(location);
    setLocations(results);
    console.log("state 11", results);
  };

  const searchWeather = (item: Location) => {
    setLocation(item)
    NavigationService.back()
  };

  const renderItem = ({item}: {item: Location}) => {
    return (
      <TouchableOpacity testID='locationItem' onPress={() => searchWeather(item)}>
        <Text style={styles.item}>
          {item?.name}, {item?.country}
        </Text>
      </TouchableOpacity>
    );
  };
  return (
    <View style={styles.container}>
      <SearchBar testId="searchInput" onSearch={handleSearch} />
      <FlatList
      testID='locationList'
        data={locations}
        keyExtractor={(item: Location) => item.id.toString()}
        renderItem={renderItem}
      />
    </View>
  );
};
const styles = StyleSheet.create({
  container: {padding: 16, flex: 1},
  item: {padding: 8, fontSize: 18, color: '#000'},
});

export default SearchScreen;
