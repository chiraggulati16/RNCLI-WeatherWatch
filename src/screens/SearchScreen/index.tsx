import {FlatList, Text, TouchableOpacity, View} from "react-native";
import {WeatherService} from "../../api/WeatherService";
import {Location} from "../../models/Location";
import SearchBar from "../../components/SearchBar";
import { useState} from "react";
import {useLocationContext} from "../../context/LocationContext";
import NavigationService from "../../navigation/NavigationService";
import { styles } from "./styles";

const SearchScreen = () => {
  const [locations, setLocations] = useState<Location[]>([]);
  const {setLocation} = useLocationContext();

  const handleSearch = async (location: string) => {
    const results = await WeatherService.getSearchList(location);
    setLocations(results);
  };

  const searchWeather = (item: Location) => {
    setLocation(item);
    NavigationService.back();
  };

  const renderItem = ({item}: {item: Location}) => {
    return (
      <TouchableOpacity
        testID="locationItem"
        onPress={() => searchWeather(item)}>
        <Text style={styles.item}>
          {item?.name}, {item?.country}
        </Text>
      </TouchableOpacity>
    );
  };
  return (
    <View style={styles.container}>
      <SearchBar testId="searchInput" onSearch={handleSearch} />
      <FlatList
        testID="locationList"
        data={locations}
        keyExtractor={(item: Location) => item.id.toString()}
        renderItem={renderItem}
      />
    </View>
  );
};

export default SearchScreen;
