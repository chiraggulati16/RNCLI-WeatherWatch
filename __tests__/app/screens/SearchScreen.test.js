import * as React from "react";
import { mount, shallow } from "enzyme";
import SearchScreen from "../../../src/screens/SearchScreen";
import { LocationContext } from '../../../src/context/LocationContext';
import { renderWithContext } from "../../mountWithContext";
import { toJson} from "enzyme-to-json"
import { WeatherService } from "../../../src/api/WeatherService";
import { act } from 'react-dom/test-utils';
import NavigationService from "../../../src/navigation/NavigationService";
let wrapper;

const defaultProps = {
    location : {
        id: 1,
        name: "Chandigarh",
        latitude: 30.7333,
        longitude: 76.7794,
        country: "India"
      },
      setLocation : jest.fn()
}
afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
})
jest.mock('../../../src/api/WeatherService');

describe("<SearchScreen/> Unit test cases using fully mounted SearchScreen Component", ()=> {

    it("<SearchScreen/> generate snapshot", ()=> {
        wrapper = renderWithContext(<SearchScreen/>);
        expect(wrapper).toMatchSnapshot();
    })

    it("<SearchScreen/> render without crashing", ()=> {
        wrapper = renderWithContext(<SearchScreen/>);
        expect(wrapper).toBeTruthy();
    })

    const mockWeatherService = WeatherService ;
    const mockResults = [
        {
            id: 2,
            name: "Delhi",
            latitude: 28.7041,
            longitude: 77.1025,
            country: "India"
          }
    ]
    beforeEach(() => {
        jest.clearAllMocks();
        mockWeatherService.getSearchList.mockResolvedValue(mockResults);
      });
    it("<SearchScreen/> handle onSearch", async ()=> {
        
        wrapper = renderWithContext(<SearchScreen/>);
        const searchInput = wrapper.find({testID: "searchInput"}).first();
        searchInput.props().onChangeText("Delhi");
        wrapper.setProps({});
        wrapper.update();
        const searchInputUpdated = wrapper.find({testID: "searchInput"}).first();
        expect(searchInputUpdated.props().value).toBe("Delhi");
       
        const searchBtn = wrapper.find({testID: "searchBtn"}).first();
        searchBtn.props().onPress();
        
        await act(async () => {
            // Wait for any promises to resolve
            await Promise.resolve();
            wrapper.setProps({})
            wrapper.update();
          });
        const locationList = wrapper.find({testID: "locationList"}).first();
        expect(locationList.props().data).toHaveLength(1);
    })

    it("<SearchScreen/> handle flatlist click", async ()=> {
        NavigationService.back = jest.fn();
        wrapper = renderWithContext(<SearchScreen/>);
        const searchInput = wrapper.find({testID: "searchInput"}).first();
        searchInput.props().onChangeText("Delhi");
        wrapper.setProps({});
        wrapper.update();
       
        const searchBtn = wrapper.find({testID: "searchBtn"}).first();
        searchBtn.props().onPress();
        
        await act(async () => {
            // Wait for any promises to resolve
            await Promise.resolve();
            wrapper.setProps({})
            wrapper.update();
          });
        const locationList = wrapper.find({testID: "locationList"}).first();
        const renderItem = renderWithContext(locationList.props().renderItem({item : mockResults[0]}))
        expect(renderItem).toBeTruthy();
        
        const locationName = renderItem.find({testID: "locationItem"}).first();
        locationName.props().onPress();
        expect(NavigationService.back).toHaveBeenCalledTimes(1);
    })
})