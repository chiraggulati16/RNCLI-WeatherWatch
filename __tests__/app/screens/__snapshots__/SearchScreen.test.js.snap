// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchScreen/> Unit test cases using fully mounted SearchScreen Component <SearchScreen/> generate snapshot 1`] = `ReactWrapper {}`;

exports[`<SearchScreen/> Unit test cases using fully mounted SearchScreen Component <SearchScreen/> render without crashing 1`] = `ReactWrapper {}`;
